sycl/test/regression/unnamed-lambda.cpp

```diff
@@ -33,3 +33,15 @@ void bar(cl::sycl::queue queue) {
     cgh.parallel_for<class K2>(cl::sycl::range<2>{1024,768}, f);
   });
 }
+
+// This validates the case where using a lambda in a kernel in a different order
+// than the lexical order of the lambdas. In a previous implementation of
+// __builtin_sycl_unique_stable_name this would result in the value of the
+// builtin being invalidated, causing a compile error. The redesigned
+// implementation should no longer have a problem with this pattern.
+void out_of_order_decls() {
+  auto w = [](auto i) {};
+  sycl::queue q;
+  q.parallel_for(10, [](auto i) {});
+  q.parallel_for(10, w);
+}
```

Comment:
Do you think there would be value in adding a test that uses macros in a problematic way to demonstrate that we don't assert/crash/do bad things? Or do you think we have sufficient coverage for that already?

Reply:
I don't see value in seeing if the compiler would crash/etc.  But we DO know that the runtime behavior of those programs would be bad.  I have no idea how to test those situations.

Reply:
I don't want to see if we do crash; I want to prove that we don't crash in a known-problematic situation.

Perhaps instead of a FE test, it'd make sense for a SYCL end-to-end test? I don't want to codify "this is the behavior you can rely on when you do this stuff", but I do think there's some benefit to showing "we know this code pattern is a sharp edge and we explicitly don't care what happens so long as 'compiler crash' isn't the behavior."

Reply:
I see.  The expected outcome is that the 'wrong' kernel would be called, or the runtime would fail (depending on how this is configured).  We're intentionally making this "UB" in the standard to cover both of those cases.

Would it be ok if @AlexeySachkov or @romanovvlad could write a test in a separate patch?

Reply:
It'd be totally fine to add a test in a separate review; no need to block this review on it. I agree on the expected outcome -- the value I was seeing in the test is showing that we don't care about the runtime behavior (you get what you get), but that we don't expect the FE to explode on it. However, if others don't think that's a particularly valuable test, I don't insist on one being added at all.

Reply:
Not sure if it makes a lot of sense to test UB. AFAIK users do not look at our tests to see what is code is legal.

Reply:
The goal isn't to "test UB", it's to *ensure our product doesn't crash just because of UB*. We added new assertions, we changed old ones, and we did it on a fast-turnaround patch on something we're now on our third iteration of due to issues discovered in the field. As I said, I don't insist on the test, but hopefully we're at least all talking about the same thing being tested when we make that decision.